py/asmrv32.c

```diff
@@ -0,0 +1,761 @@
+
+// wrapper around everything in this file
+#if MICROPY_EMIT_RV32
+
+#include "py/asmrv32.h"
+
+#if MICROPY_DEBUG_VERBOSE
+#define DEBUG_PRINT (1)
+#define DEBUG_printf DEBUG_printf
+#else
+#define DEBUG_printf(...) (void)0
+#endif
+
+#ifndef MP_POPCOUNT
+#ifdef _MSC_VER
+#include <intrin.h>
+#define MP_POPCOUNT __popcnt
+#else
+#if defined __has_builtin
+#if __has_builtin(__builtin_popcount)
+#define MP_POPCOUNT __builtin_popcount
+#endif
+#else
+static uint32_t fallback_popcount(uint32_t value) {
+    value = value - ((value >> 1) & 0x55555555);
+    value = (value & 0x33333333) + ((value >> 2) & 0x33333333);
+    value = (value + (value >> 4)) & 0x0F0F0F0F;
+    return value * 0x01010101;
+}
+#define MP_POPCOUNT fallback_popcount
+#endif
+#endif
+#endif
+
+#define INTERNAL_TEMPORARY ASM_RV32_REG_S1
+#define AVAILABLE_REGISTERS_COUNT 32
+#define FITS_IN_12_BITS_SIGNED(value) (((value) >= -2048) && ((value) <= 2047))
+
+#define EMIT_NOP(state)                                                        \
+    asm_rv32_opcode_addi(state, ASM_RV32_REG_ZERO, ASM_RV32_REG_ZERO, 0);
+
+///////////////////////////////////////////////////////////////////////////////
+
+#define RV32_ENCODE_TYPE_B(op, ft3, rs1, rs2, imm)                                              \
+    ((op & 0b1111111) | ((ft3 & 0b111) << 12) | ((imm & 0b100000000000) >> 4) |  \
+    ((imm & 0b11110) << 7) | ((rs1 & 0b11111) << 15) |                          \
+    ((rs2 & 0b11111) << 20) | ((imm & 0b11111100000) << 20) |                   \
+    ((imm & 0b1000000000000) << 19))
+
+#define RV32_ENCODE_TYPE_I(op, ft3, rd, rs, imm)                               \
+    ((op & 0b1111111) | ((rd & 0b11111) << 7) | ((ft3 & 0b111) << 12) |          \
+    ((rs & 0b11111) << 15) | ((imm & 0b111111111111) << 20))
+
+#define RV32_ENCODE_TYPE_J(op, rd, imm)                                        \
+    ((op & 0b1111111) | ((rd & 0b11111) << 7) | (imm & 0b11111111000000000000) | \
+    ((imm & 0b100000000000) << 9) | ((imm & 0b11111111110) << 20) |             \
+    ((imm & 0b100000000000000000000) << 11))
+
+#define RV32_ENCODE_TYPE_R(op, ft3, ft7, rd, rs1, rs2)                         \
+    ((op & 0b1111111) | ((rd & 0b11111) << 7) | ((ft3 & 0b111) << 12) |          \
+    ((rs1 & 0b11111) << 15) | ((rs2 & 0b11111) << 20) |                         \
+    ((ft7 & 0b1111111) << 25))
+
+#define RV32_ENCODE_TYPE_S(op, ft3, rs1, rs2, imm)                             \
+    ((op & 0b1111111) | ((imm & 0b11111) << 7) | ((ft3 & 0b111) << 12) |         \
+    ((rs1 & 0b11111) << 15) | ((rs2 & 0b11111) << 20) |                         \
+    ((imm & 0b111111100000) << 20))
+
+#define RV32_ENCODE_TYPE_U(op, rd, imm)                                        \
+    ((op & 0b1111111) | ((rd & 0b11111) << 7) |                                  \
+    (imm & 0b11111111111111111111000000000000))
+
+// ADD RD, RS1, RS2
+void asm_rv32_opcode_add(asm_rv32_t *state, mp_uint_t rd, mp_uint_t rs1, mp_uint_t rs2) {
```

Comment:
I would think that all of these functions could be made `static inline` in the header file, and it would result in smaller code.  You can see that other emitters like `asmxtensa.h` have them inline.

But that would only be true if the `saved_registers_mask` was either removed (and a different way found to work out which registers to save), or made more general, eg the `asm_rv32_emit_instruction()` function may be able to tell which register is written to in a general way, and then mark it as used.

Reply:
Luckily RV32 opcodes are quite regular when it comes to register locations within opcodes and their encoding.  To remove the register usage tracking in the opcode functions I can have something like `asm_rv32_emit_instruction(asm_rv32_t *emit, uint32_t opcode, rv32_opcode_kind_t kind)` where the latter is just an enumeration indicating which encoding to follow.

Reply:
The tracking was removed in 4d0c3f964c2f8fb6bb3d5eaff95b989bf6e1e7f8 so this doesn't apply anymore.